modules/dxp/apps/saml/saml-persistence-service/src/main/java/com/liferay/saml/persistence/internal/upgrade/v3_0_0/SamlSpSessionUpgradeProcess.java

```diff
@@ -38,35 +39,54 @@ protected void doUpgrade() throws Exception {
 					new AlterTableAddColumn("samlPeerBindingId", "LONG null"));
 			}
 
-			int samlSpSessionIdOffset = 0;
+			int samlSpSessionIdOffset = _getSamlSpSessionIdOffset();
+
+			int latestSamlPeerBindingId = _getLatestSamlPeerBindingId();
 
 			try (PreparedStatement preparedStatement =
 					connection.prepareStatement(
-						"select min(samlSpSessionId) - 1 from SamlSpSession");
+						StringBundler.concat(
+							"select min(samlIdpSpSessionId) as ",
+							"samlIdpSpSessionId, companyId, min(createDate) ",
+							"as createDate, userId, userName, nameIdFormat, ",
+							"nameIdNameQualifier, nameIdValue, samlSpEntityId ",
+							"from SamlIdpSpSession group by companyId, ",
+							"userId, userName, samlSpEntityId, nameIdFormat, ",
+							"nameIdValue"));
 				ResultSet resultSet = preparedStatement.executeQuery()) {
 
-				if (resultSet.next()) {
-					samlSpSessionIdOffset = resultSet.getInt(1);
+				while (resultSet.next()) {
+					int samlIdpSpSessionId = resultSet.getInt(
+						"samlIdpSpSessionId");
+					long companyId = resultSet.getLong("companyId");
+					Timestamp createDate = resultSet.getTimestamp("createDate");
+					long userId = resultSet.getLong("userId");
+					String userName = resultSet.getString("userName");
+					String nameIdFormat = resultSet.getString("nameIdFormat");
+					String nameIdNameQualifier = resultSet.getString(
+						"nameIdNameQualifier");
+					String nameIdValue = resultSet.getString("nameIdValue");
+					String samlSpEntityId = resultSet.getString(
+						"samlSpEntityId");
+
+					int samlPeerBindingId =
+						samlIdpSpSessionId + -samlSpSessionIdOffset +
+							latestSamlPeerBindingId;
+
+					runSQL(
```

Comment:
The runSQL() method will commit after each SQL statement, so if there is a `SQLException`, earlier SQL will re-run and cause duplicate entries in the `SamlPeerBinding` table. I think we can work around this by using its overloaded variety which accepts an existing `Connection`.

Reply:
Done, take a look